internal/template-validator/tlsinfo/tlsinfo.go

```diff
@@ -3,150 +3,138 @@ package tlsinfo
 
-	"github.com/fsnotify/fsnotify"
-
 	"kubevirt.io/ssp-operator/internal/common"
+	"kubevirt.io/ssp-operator/internal/template-validator/filewatch"
 	"kubevirt.io/ssp-operator/internal/template-validator/logger"
 )
 
 const (
-	CertFilename         = "tls.crt"
-	KeyFilename          = "tls.key"
-	retryInterval        = 1 * time.Minute
+	CertFilename       = "tls.crt"
+	KeyFilename        = "tls.key"
+	TLSOptionsFilename = "tls-config.json"
+
 	CiphersEnvName       = "TLS_CIPHERS"
 	TLSMinVersionEnvName = "TLS_MIN_VERSION"
 )
 
 type TLSInfo struct {
-	CertsDirectory string
-	cert           *tls.Certificate
-	certLock       sync.Mutex
-	stopCertReload chan struct{}
-	sspTLSOptions  common.SSPTLSOptions
-}
+	CertsDirectory      string
+	TLSOptionsDirectory string
 
-func (ti *TLSInfo) IsEnabled() bool {
-	return ti.CertsDirectory != ""
-}
+	lock sync.Mutex
 
-func (ti *TLSInfo) Init() {
-	ti.InitCerts()
-	ti.InitCryptoConfig()
-}
+	watch     filewatch.Watch
+	stopWatch chan struct{}
 
-func (ti *TLSInfo) InitCryptoConfig() {
-	nonSplitCiphers, _ := os.LookupEnv(CiphersEnvName)
-	ti.sspTLSOptions.OpenSSLCipherNames = strings.Split(nonSplitCiphers, ",")
-	ti.sspTLSOptions.MinTLSVersion, _ = os.LookupEnv(TLSMinVersionEnvName)
+	cert      *tls.Certificate
+	certError error
+
+	cipherSuites    []uint16
+	minTLSVersion   uint16
+	tlsOptionsError error
 }
 
-func (ti *TLSInfo) InitCerts() {
-	if !ti.IsEnabled() {
-		return
+func (ti *TLSInfo) Init() error {
+	if ti.stopWatch != nil {
+		return fmt.Errorf("certificate watcher was already initialized")
 	}
 
-	directory := ti.CertsDirectory
-	filesChanged, watcherCloser, err := watchDirectory(directory)
-	if err != nil {
-		panic(err)
+	ti.watch = filewatch.New()
+
+	if err := ti.initCerts(); err != nil {
+		return err
 	}
 
-	ti.stopCertReload = make(chan struct{})
-	notify(filesChanged)
+	if err := ti.initCryptoConfig(); err != nil {
+		return err
+	}
 
+	ti.stopWatch = make(chan struct{})
 	go func() {
-		defer watcherCloser.Close()
 		for {
 			select {
-			case <-filesChanged:
-				err := ti.updateCertificates(directory)
-				if err != nil {
-					go func() {
-						time.Sleep(retryInterval)
-						notify(filesChanged)
-					}()
-				}
-			case <-ti.stopCertReload:
+			case <-ti.stopWatch:
+				return
+			default:
+			}
+
+			watchErr := ti.watch.Run(ti.stopWatch)
+			if watchErr == nil {
+				return
+			}
+			// Log error and restart watch
+			logger.Log.Info("failed watching files", "error", watchErr)
+
+			select {
+			case <-ti.stopWatch:
 				return
+			case <-time.After(time.Second):
 			}
 		}
 	}()
-}
 
-func (ti *TLSInfo) Clean() {
-	if ti.stopCertReload != nil {
-		close(ti.stopCertReload)
-	}
+	return nil
 }
 
-func watchDirectory(directory string) (chan struct{}, io.Closer, error) {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		logger.Log.Error(err, "Failed to create an inotify watcher")
-		return nil, nil, err
-	}
+const watchErrorFmt = "error adding directory to watch: %w"
```

Comment:
Personal preference, but move it to the beginning of the file?

Reply:
I prefer it to be close the where it is used.